waku/v2/peermanager/topic_event_handler.go

```diff
@@ -0,0 +1,167 @@
+		if pm.host.Network().Connectedness(peer) == network.Connected {
+			connectedPeers++
+		} else {
+			notConnectedPeers = append(notConnectedPeers, peer)
+		}
+	}
+
+	if connectedPeers >= relayOptimalPeersPerShard { //TODO: Use a config rather than hard-coding.
+		// Should we use optimal number or define some sort of a config for the node to choose from?
+		// A desktop node may choose this to be 4-6, whereas a service node may choose this to be 8-12 based on resources it has
+		// or bandwidth it can support.
+		// Should we link this to bandwidth management somehow or just depend on some sort of config profile?
+		pm.logger.Info("Optimal required relay peers for new pubSubTopic are already connected ", zap.String("pubSubTopic", pubsubTopic),
+			zap.Int("connectedPeerCount", connectedPeers))
+		return
+	}
+	triggerDiscovery := false
+	if notConnectedPeers.Len() > 0 {
+		numPeersToConnect := notConnectedPeers.Len() - connectedPeers
+		if numPeersToConnect < 0 {
+			numPeersToConnect = notConnectedPeers.Len()
+		} else if numPeersToConnect-connectedPeers > relayOptimalPeersPerShard {
+			numPeersToConnect = relayOptimalPeersPerShard - connectedPeers
+		}
+		if numPeersToConnect+connectedPeers < relayOptimalPeersPerShard {
+			triggerDiscovery = true
+		}
+		//For now all peers are being given same priority,
+		// Later we may want to choose peers that have more shards in common over others.
+		pm.connectToPeers(notConnectedPeers[0:numPeersToConnect])
+	} else {
+		triggerDiscovery = true
+	}
+
+	if triggerDiscovery {
+		//TODO: Initiate on-demand discovery for this pubSubTopic.
+		// Use peer-exchange and rendevouz?
+		//Should we query discoverycache to find out if there are any more peers before triggering discovery?
+		return
+	}
+}
+
+func (pm *PeerManager) handleNewRelayTopicUnSubscription(pubsubTopic string) {
+	pm.logger.Info("handleNewRelayTopicUnSubscription", zap.String("pubSubTopic", pubsubTopic))
+	pm.topicMutex.Lock()
+	defer pm.topicMutex.Unlock()
+	_, ok := pm.subRelayTopics[pubsubTopic]
+	if !ok {
+		//Nothing to be done, as we are already unsubscribed from this topic.
+		return
+	}
+	delete(pm.subRelayTopics, pubsubTopic)
+
+	//If there are peers only subscribed to this topic, disconnect them.
+	relevantPeersForPubSubTopic := pm.host.Peerstore().(*wps.WakuPeerstoreImpl).PeersByPubSubTopic(pubsubTopic)
+	for _, peer := range relevantPeersForPubSubTopic {
+		if pm.host.Network().Connectedness(peer) == network.Connected {
+			peerTopics, err := pm.host.Peerstore().(*wps.WakuPeerstoreImpl).PubSubTopics(peer)
+			if err != nil {
+				pm.logger.Error("Could not retrieve pubsub topics for peer", zap.Error(err),
+					logging.HostID("peerID", peer))
+				continue
+			}
+			if len(peerTopics) == 1 && peerTopics[0] == pubsubTopic {
+				err := pm.host.Network().ClosePeer(peer)
+				if err != nil {
+					pm.logger.Warn("Failed to disconnect connection towards peer",
+						logging.HostID("peerID", peer))
+				}
```

Comment:
Should `continue` here?
```suggestion
				if err != nil {
					pm.logger.Warn("Failed to disconnect connection towards peer",
						logging.HostID("peerID", peer))
					continue
				}
```